lib/adapters/JasmineAdapter.js

```diff
@@ -4,68 +4,115 @@ import {Test, Suite} from "../Data.js";
+            }
+        }
+        var suite = new Suite(name, childSuites, tests);
+
+        this.idObjectMap[jasmineSuite.id] = suite;
 
-    onSpecStarted() {
-        this.startTime = new Date();
+        return suite;
     }
 
-    onSpecDone(details) {
-        var runtime = new Date() - this.startTime;
-        var testName = details.description;
-        var suiteName = details.fullName;
-        var errors = details.failedExpectations;
-        var status;
-        if (details.status === "pending") {
-            status = "skipped";
-        } else {
-            status = details.status;
-        }
+    /**
+     * Converts a jasmine spec into a standardized test.
+     */
+    convertSpec(jasmineSpec) {
+        var testName = jasmineSpec.description;
+        var suiteName = jasmineSpec.result.fullName;
 
         // Jasmine only provides a concatenated suiteName + testName string.
         // Remove the testName from the string.
         if (suiteName.lastIndexOf(testName) === suiteName.length - testName.length) {
             suiteName = suiteName.substr(0, suiteName.length - testName.length - 1);
         }
+
         var test = new Test(
             testName,
             suiteName,
-            status,
-            runtime,
-            errors
+            undefined,
+            undefined,
+            undefined
         );
+
+        this.idObjectMap[jasmineSpec.id] = test;
+
+        return test;
+    }
+
+
+    onJasmineStarted() {
+        var startSuite = this.convertSuite(this.jasmine.topSuite());
+        startSuite.name = "";
+        this.idObjectMap.globalSuite = startSuite;
+
+        this.emit("runStart", startSuite);
+    }
+
+
+    onSpecStarted(details) {
+        this.startTime = new Date();
+        this.emit("testStart", this.idObjectMap[details.id]);
+    }
+
+    onSpecDone(details) {
+        var test = this.idObjectMap[details.id];
+        test.runtime = new Date() - this.startTime;
+        test.errors = details.failedExpectations;
+        if (details.status === "pending") {
+            test.status = "skipped";
+        } else {
+            test.status = details.status;
+        }
+
         this.emit("testEnd", test);
-        this.tests[this.tests.length - 1].push(test);
     }
 
-    onSuiteStarted() {
-        this.tests.push([]);
-        this.suites.push([]);
+    onSuiteStarted(details) {
+        this.emit("suiteStart", this.idObjectMap[details.id]);
     }
 
     onSuiteDone(details) {
-        if (details.failedExpectations.length > 0) {
-            this.tests[this.tests.length - 1].push(new Test("afterAll", "failed", 0));
-        }
-        var suite = new Suite(details.description, this.suites.pop(), this.tests.pop());
-        this.emit("suiteEnd", suite);
-        this.suites[this.suites.length - 1].push(suite);
+
+        this.emit("suiteEnd", this.idObjectMap[details.id]);
+
+        /*
+         FIXME: Jasmine Suite may have failedExpectations caused by a failure in afterAll.
```

Comment:
This seems fair to ignore in the adapter, but document it as a limitation and point it out to the Jasmine team.